### PR TITLE
StarterPackMgr

### PR DESCRIPTION
**Pull request by AlexApps99**
This change is [Reviewable](https://reviewable.io/reviews/zeldaret/botw/60)

### CODE REVIEW

**Review by leoetlino — COMMENTED** [review]
Looks good so far :+1: 

> Reviewed 8 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/60#-:-MhDP71MANIJjrv5BErQ:b-n0beyd)* status: all files reviewed, 1 unresolved discussion (waiting on @AlexApps99)

___
*[src/KingSystem/System/StarterPackMgr.cpp, line 3 at r1](https://reviewable.io/reviews/zeldaret/botw/60#-MhDOz5GE1YrAcj5ysFH:-MhDOz5GE1YrAcj5ysFI:b-kxqdt9) ([raw file](https://github.com/zeldaret/botw/blob/b40be97e8c4239b82dd4a6f18c6cee3896a8671b/src/KingSystem/System/StarterPackMgr.cpp#L3)):*
> ```cpp
> #include "KingSystem/System/StarterPackMgr.h"
> 
> #define WARCRIME(cmp) (GameConfig::getInstance()->mField419 cmp 0)
> ```

don't forget to remove this

File: data/uking_functions.csv [hunk 1]
```diff
@@ -53048,12 +53048,12 @@ Address,Quality,Size,Name
...
 0x00000071008a5b1c,U,000068,
 0x00000071008a5b60,U,000008,DungeonPackMgr::startedLoadAndFailed
-0x00000071008a5b68,U,000076,EntryFactorySARC::dtorDelete
-0x00000071008a5bb4,U,000132,EntryFactorySARC::rtti1
-0x00000071008a5c38,U,000092,EntryFactorySARC::rtti2
-0x00000071008a5c94,U,000152,EntryFactorySARC::newResource
-0x00000071008a5d2c,U,000008,EntryFactorySARC::getResourceSize
-0x00000071008a5d34,U,000008,EntryFactorySARC::getLoadDataAlignment
+0x00000071008a5b68,O,000076,_ZN4ksys3res12EntryFactoryINS0_7ArchiveEED0Ev
+0x00000071008a5bb4,M,000132,_ZNK4ksys3res12EntryFactoryINS0_7ArchiveEE27checkDerivedRuntimeTypeInfoEPKN4sead15RuntimeTypeInfo9InterfaceE
```
Comment (AlexApps99) [inline data/uking_functions.csv:53052]: Not sure if this should be a major or minor diff

File: src/KingSystem/Resource/resResourceMgrTask.h [hunk 2]
```diff
@@ -264,6 +265,10 @@ class ResourceMgrTask : public sead::CalculateTask, public sead::hostio::Node {
     ResourceUnitDelegatePair& getUnitInitLoadFn() { return mUnitInitLoadFn; }
     auto& getUnitAdjustHeapFn() { return mUnitAdjustHeapFn; }
 
+    OverlayArena* getArenaForResourceL() { return &mArenaForResourceL; }
+
+    void setPack(Handle*);
```
Comment (AlexApps99) [inline src/KingSystem/Resource/resResourceMgrTask.h:270]: Should it have the name of the parameter?

File: src/KingSystem/Resource/resSystem.cpp [hunk 3]
```diff
@@ -36,4 +36,12 @@ s32 getDefaultAlignment() {
     return 8;
 }
 
+void registerPackExtension(bool, const sead::SafeString&) {}
```
Comment (AlexApps99) [inline src/KingSystem/Resource/resSystem.cpp:39]: Should it have the name of the parameter?

File: src/KingSystem/Resource/resSystem.cpp [hunk 4]
```diff
@@ -36,4 +36,12 @@ s32 getDefaultAlignment() {
...
+
+void stubbedBool(bool) {}
```
Comment (AlexApps99) [inline src/KingSystem/Resource/resSystem.cpp:41]: Should it have the name of the parameter?

File: src/KingSystem/Resource/resSystem.h [hunk 5]
```diff
@@ -27,4 +28,14 @@ bool returnFalse3(const sead::SafeString& path);
 
 s32 getDefaultAlignment();
 
+// In release builds, this function does nothing.
+// TODO: figure out what this is used for. Stubbed log function?
+void registerPackExtension(bool, const sead::SafeString&);
```
Comment (AlexApps99) [inline src/KingSystem/Resource/resSystem.h:33]: Should it have the name of the parameter?

File: src/KingSystem/Resource/resSystem.h [hunk 6]
```diff
@@ -27,4 +28,14 @@ bool returnFalse3(const sead::SafeString& path);
...
+
+// In release builds, this function does nothing.
+// TODO: figure out what this is used for. Stubbed log function?
+void stubbedBool(bool);
```
Comment (AlexApps99) [inline src/KingSystem/Resource/resSystem.h:37]: Should it have the name of the parameter?

File: src/KingSystem/Resource/resSystem.h [hunk 7]
```diff
@@ -27,4 +28,14 @@ bool returnFalse3(const sead::SafeString& path);
...
+
+void setResourceMgrPack(Handle*);
```
Comment (AlexApps99) [inline src/KingSystem/Resource/resSystem.h:39]: Should it have the name of the parameter?

File: src/KingSystem/System/StarterPackMgr.cpp [hunk 8]
```diff
@@ -0,0 +1,220 @@
+#include "KingSystem/System/StarterPackMgr.h"
+
+// TODO the LoadRequest stuff would be nicer as an inline function
+#define LOAD_REQ()                                                                                 \
```
Comment (AlexApps99) [inline src/KingSystem/System/StarterPackMgr.cpp:4]: Should I ditch the macro and put it in each function individually?

Comment (AlexApps99) [reply src/KingSystem/System/StarterPackMgr.cpp:4]: I also need to get rid of the comment with the TODO

File: src/KingSystem/System/StarterPackMgr.cpp [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+void StarterPackMgr::unloadTitleBGAndAocMainFieldPacksAndStuff() {
+    if (!GameConfig::getInstance()->mField419) {
+        res::setResourceMgrPack(nullptr);
+        mTitleBGPack.requestUnload();
+        mTitleBGLangPack.requestUnload();
+
+        if (uking::aoc::Manager::instance())
+            uking::aoc::Manager::instance()->unloadAocMainFieldPack();
+
+        for (unsigned char x = 0; x < 5; x++)
```
Comment (AlexApps99) [inline src/KingSystem/System/StarterPackMgr.cpp:197]: Should there be a constant for this?

File: src/KingSystem/System/StarterPackMgr.h [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace ksys {
+
+class StarterPackMgr {
+    SEAD_SINGLETON_DISPOSER(StarterPackMgr)
+    StarterPackMgr();
+    ~StarterPackMgr();
+
+public:
+    void registerPackFactoryAndMakeOverlayArena(bool);
```
Comment (AlexApps99) [inline src/KingSystem/System/StarterPackMgr.h:25]: Should it have the name of the parameter?

**Review by AlexApps99 — COMMENTED** [review]
Some things that might need to be addressed

File: data/uking_functions.csv (hunk 1 above)
Comment (AlexApps99) [reply data/uking_functions.csv:53052]: I changed this to minor in the latest commit, but still not 100% sure

Comment (AlexApps99) [reply data/uking_functions.csv:53052]: I ended up changing it to minor but still not 100% sure

**Review by leoetlino — COMMENTED** [review]
> Reviewed 3 of 8 files at r2, 5 of 5 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/60)* status: all files reviewed, 11 unresolved discussions (waiting on @AlexApps99)

___
*[data/uking_functions.csv, line 53052 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhDNn5JCALJdapO4sZ5-r2-53052:-MhYBH38AUJukbTrUb79:b-qwh1qe) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/data/uking_functions.csv#L53052)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

I changed this to minor in the latest commit, but still not 100% sure
</blockquote></details>

Minor is fine, since we're completely sure that the function matches.
___
*[src/KingSystem/Resource/resResourceMgrTask.h, line 270 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhDNn5JCALJdapO4sZ9-r2-270:-MhYBSxY30NDbcf9k9ah:bb77ufg) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resResourceMgrTask.h#L270)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should it have the name of the parameter?
</blockquote></details>

Yes.
___
*[src/KingSystem/Resource/resSystem.h, line 33 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutm-r2-33:-MhYB_jh3xzkFWjBZuJU:bh23gbw) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.h#L33)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should it have the name of the parameter?
</blockquote></details>

Yes. Naming the bool will be a bit tricky though, so feel free to leave that one unnamed.
___
*[src/KingSystem/Resource/resSystem.h, line 37 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutm-r2-37:-MhYBgOU6teDgEgjiEeR:b-vz2a5i) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.h#L37)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should it have the name of the parameter?
</blockquote></details>

This one is tricky to name, so you can leave it unnamed if you want.
___
*[src/KingSystem/Resource/resSystem.h, line 39 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutm-r2-39:-MhYBnBWCfmUVPeEQOlw:bb77ufg) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.h#L39)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should it have the name of the parameter?
</blockquote></details>

Yes.
___
*[src/KingSystem/Resource/resSystem.cpp, line 39 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutl-r2-39:-MhYBqVz3F9GZkWPTOyL:bvg8w0s) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.cpp#L39)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should it have the name of the parameter?
</blockquote></details>

Same as for the declaration in the header.
___
*[src/KingSystem/Resource/resSystem.cpp, line 41 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutl-r2-41:-MhYBun92XTN0t1Ex7vE:bvg8w0s) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.cpp#L41)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should it have the name of the parameter?
</blockquote></details>

Same as for the declaration in the header.
___
*[src/KingSystem/System/StarterPackMgr.h, line 25 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhDNn5JCALJdapO4sZC-r2-25:-MhYCqHNEsCHkMk3CXMh:bb77ufg) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/System/StarterPackMgr.h#L25)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should it have the name of the parameter?
</blockquote></details>

Yes.
___
*[src/KingSystem/System/StarterPackMgr.h, line 25 at r3](https://reviewable.io/reviews/zeldaret/botw/60#-MhYC-h_0cBGowlByZvF:-MhYC-h_0cBGowlByZvG:b50zyrs) ([raw file](https://github.com/zeldaret/botw/blob/eb7ac5cf62955276454337d4d4089e19021ffb04/src/KingSystem/System/StarterPackMgr.h#L25)):*
> ```objc
> 
> public:
>     void registerPackFactoryAndMakeOverlayArena(bool reducedHeapSize);
> ```

`reduced_heap_size`
___
*[src/KingSystem/System/StarterPackMgr.cpp, line 4 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhDNn5JCALJdapO4sZB-r2-4:-MhYC5q6D4Uej-iuoFZT:b-w62qmu) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/System/StarterPackMgr.cpp#L4)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should I ditch the macro and put it in each function individually?
</blockquote></details>

IMO yes, it'd be clearer to remove the macro (since it declares a variable).
___
*[src/KingSystem/System/StarterPackMgr.cpp, line 197 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhDNn5JCALJdapO4sZB-r2-197:-MhYC_sj9BZ-hwjYE0g8:b-93vrdi) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/System/StarterPackMgr.cpp#L197)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Should there be a constant for this?
</blockquote></details>

Not sure it's worth making a named constant for this. By the way, does this match if you use `int` instead of `unsigned char`?

**Review by AlexApps99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/60)* status: all files reviewed, 11 unresolved discussions (waiting on @AlexApps99 and @leoetlino)

___
*[src/KingSystem/Resource/resSystem.h, line 33 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutm-r2-33:-MhYGlQDDheIDSthFqzm:b-ckjo8v) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.h#L33)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

Yes. Naming the bool will be a bit tricky though, so feel free to leave that one unnamed.
</blockquote></details>

Since all `true`s had ".pack" and all `false`s had "", I think has_extension might be a good one
___
*[src/KingSystem/Resource/resSystem.h, line 37 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutm-r2-37:-MhYFwhdAY8k03ITC0yD:bmk34r3) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.h#L37)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

This one is tricky to name, so you can leave it unnamed if you want.
</blockquote></details>

It's only used once, so I don't think it would be possible to name

**Review by leoetlino — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/60)* status: all files reviewed, 11 unresolved discussions (waiting on @AlexApps99)

___
*[src/KingSystem/Resource/resSystem.h, line 33 at r2](https://reviewable.io/reviews/zeldaret/botw/60#-MhIa8clmfRWUsTMGutm-r2-33:-MhYHCjJ0yW2gcDnUFPj:b-8fx2u7) ([raw file](https://github.com/zeldaret/botw/blob/d99d8c0a495e6cb8c56c49671ec96c52c0737468/src/KingSystem/Resource/resSystem.h#L33)):*
<details><summary><i>Previously, AlexApps99 wrote…</i></summary><blockquote>

Since all `true`s had ".pack" and all `false`s had "", I think has_extension might be a good one
</blockquote></details>

:+1:

**Review by AlexApps99 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/60)* status: 7 of 11 files reviewed, 2 unresolved discussions (waiting on @AlexApps99 and @leoetlino)

___
*[src/KingSystem/System/StarterPackMgr.h, line 25 at r3](https://reviewable.io/reviews/zeldaret/botw/60#-MhYC-h_0cBGowlByZvF:-MhYI6yXEe0c249bNsKe:b-896fix) ([raw file](https://github.com/zeldaret/botw/blob/eb7ac5cf62955276454337d4d4089e19021ffb04/src/KingSystem/System/StarterPackMgr.h#L25)):*
<details><summary><i>Previously, leoetlino (Léo Lam) wrote…</i></summary><blockquote>

`reduced_heap_size`
</blockquote></details>

Done.

**Review by leoetlino — APPROVED** [review]
:lgtm: (after the commits are squashed)

> Reviewed 4 of 4 files at r4, all commit messages.
> Dismissed @AlexApps99 from a discussion.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/60#-:-MhYIY644LTwmYhgwtgg:b-8wfb01)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @AlexApps99)

**Comment by AlexApps99** [comment]
Ready to merge, thanks for all the help - I feel like I've learned a lot :+1:

**Review by AlexApps99 — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zeldaret/botw/60)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @AlexApps99)

**Pull request merged**